### PR TITLE
PyTorch: save/load models

### PR DESCRIPTION
**Pull request by patrick-wilken**
".pt" file extension seems standard.
I think calling `model.load_state_dict()` in each epoch is the right way to do it if we really want `get_model_func` to be specific to the epoch. Although this probably has to be extended if the set of model parameters really changes between epochs.

The demo now carries over the model state between model epochs, saves model files, and can be interrupted and continued.

### CODE REVIEW

File: returnn/engine/base.py [hunk 1]
```diff
@@ -56,6 +56,10 @@ def get_existing_models(cls, config):
           if os.path.exists(fn + ".index"):
             file_list[epoch] = fn
             break
+        elif util.BackendEngine.is_torch_selected():
+          if os.path.exists(fn + ".pt"):
+            file_list[epoch] = fn
+            break
```
Comment (albertz) [inline returnn/engine/base.py:62]: Why is this needed? I think this can be removed.

File: returnn/torch/engine.py [hunk 2]
```diff
@@ -52,6 +57,11 @@ def train(self):
     start_epoch, _ = self.get_train_start_epoch_batch(self.config)
     final_epoch = self.config_get_final_epoch(self.config)
 
+    if start_epoch == 1:
+      self._model_state = None
+    else:
+      self._model_state = self._load_model_state_from_file(epoch=start_epoch - 1)
```
Comment (albertz) [inline returnn/torch/engine.py:63]: Why keeping this in an attrib? I would load the params directly into the model, and only when this is really needed.

File: returnn/torch/engine.py [hunk 3]
```diff
@@ -115,6 +127,31 @@ def train_epoch(self):
...
+    """
+    :param int epoch:
+    :return: state saved via previous self._save_model_state_to_file() call for given epoch
+    :rtype: collections.OrderedDict
+    """
+    filename = self.get_epoch_model_filename(epoch=epoch) + util.get_model_filename_postfix()
+    print("Load model %s" % (filename,), file=log.v4)
+    return torch.load(filename)
+
+  def _save_model_state_to_file(self, model_state_dict):
```
Comment (albertz) [inline returnn/torch/engine.py:143]: I would keep this more consistent to the TF/Theano logic, and directly store it based on the current loaded model, not via model state dict.

File: returnn/util/basic.py [hunk 4]
```diff
@@ -314,6 +314,8 @@ def get_model_filename_postfix():
   if BackendEngine.is_tensorflow_selected():
     # There will be multiple files but a *.meta file will always be present.
     return ".meta"
+  if BackendEngine.is_torch_selected():
+    return ".pt"
```
Comment (albertz) [inline returnn/util/basic.py:318]: Why is this needed? I don't think you want that.

File: returnn/util/basic.py [hunk 5]
```diff
@@ -329,6 +331,8 @@ def get_checkpoint_filepattern(filepath):
     return filepath[:-len(".meta")]
   elif filepath.endswith(".index"):
     return filepath[:-len(".index")]
+  elif filepath.endswith(".pt"):
+    return filepath[:-len(".pt")]
```
Comment (albertz) [inline returnn/util/basic.py:335]: Remove this.

**Comment by albertz** [comment]
> ".pt" file extension seems standard.

Why does that matter? There is no point in hardcoding this. Just leave it up to the user.

And note, the TF logic is very different there. It is about the **additional** postfix to the file, after the standard file extension (which is often sth like `.ckpt` but also totally up to the user), which is added by TF because a checkpoint consists  multiple files. This extra postfix logic is only needed when you have this case, i.e. multiple files each with some additional file extension. This extra postfix is predefined by TF and cannot be changed. That's why it needs this extra logic for TF.

**Comment by albertz** [comment]
> Although this probably has to be extended if the set of model parameters really changes between epochs.

This is important. We should avoid the GPU -> CPU and CPU -> GPU transfer when it is not needed.

Maybe the interface for `get_model_func` is bad? If this is difficult to do in practice, due to the interface of `get_model_func`, we should change that interface.

**Comment by patrick-wilken** [comment]
> > ".pt" file extension seems standard.
> 
> Why does that matter? There is no point in hardcoding this. Just leave it up to the user.

With the current config options (`model`) the user cannot specify a file extension because the epoch (and possibly "pretrain") is added to the path. (i.e. `model.pt.005` instead of `model.005.pt`). If we want to reuse the existing code to create and find the paths something like what I implemented is needed. Or we could say we don't need the file extension...

Regarding the model: I don't think it is practical to change an existing `Module` instance (add parameters, replace the `forward` function) between epochs. So if we want to support that we get a new network each epoch this means we get a new `Module` instance, as it is currently implemented. (Maybe `get_model_func` could return `None` to indicate same as previous epoch? Then we could keep the instance.)
Of course, if we don't need that functionality we would create the `Module` instance once and load the parameters directly into it.

**Comment by albertz** [comment]
Ok then on the filename postfix.

Regarding the model, I was not saying at all that you should change an existing `Module` instance. Where did you read that?

Although, I think I would also leave that up to the user. E.g. there could be some `on_new_epoch(model)` callback, and the user could maybe just change some other hyper params, like dropout or do whatever.

> So if we want to support to have a new network each epoch this means we get a new `Module` instance, as it is currently implemented.

There are other ways to implement it, this is not strictly necessary. This is why I questioned whether our `get_model_func` interface should maybe be changed.

> (Maybe `get_model_func` could return `None` to indicate same as previous epoch? Then we could keep the instance.)

Yes, this is one example.

In any case, we should keep it efficient, and avoid introducing inefficiencies at all cost.

**Comment by albertz** [comment]
Regarding the proof of concept, maybe completely ignore such pretraining aspect for now, to keep it simple. So `get_model_func` would not get an `epoch` parameter.

**Comment by patrick-wilken** [comment]
This would be the version without the `epoch` parameter, so calling `get_model` once and then keep the `Module` as attribute.

File: returnn/torch/engine.py [hunk 6]
```diff
@@ -52,6 +57,8 @@ def train(self):
     start_epoch, _ = self.get_train_start_epoch_batch(self.config)
     final_epoch = self.config_get_final_epoch(self.config)
 
+    self._load_model(epoch=start_epoch)
```
Comment (albertz) [inline returnn/torch/engine.py:60]: This should not be here but in `init_train_from_config` or `init_network_from_config`.

Comment (albertz) [reply returnn/torch/engine.py:60]: Just like `start_epoch` etc. Maybe I just merge this for now.
